modules/metastorage-server/src/main/java/org/apache/ignite/internal/metastorage/server/persistence/RangeCursor.java

```diff
@@ -0,0 +1,182 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean hasNext() {
+        return it.hasNext();
+    }
+
+    /** {@inheritDoc} */
+    @Override public Entry next() {
+        return it.next();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void close() throws Exception {
+        // No-op.
+    }
+
+    /** {@inheritDoc} */
+    @NotNull
+    @Override public Iterator<Entry> iterator() {
+        return it;
+    }
+
+    /**
+     * Creates an iterator for this cursor.
+     *
+     * @return Iterator.
+     */
+    @NotNull
+    private Iterator<Entry> createIterator() {
+        return new Iterator<>() {
+            /** {@inheritDoc} */
+            @Override public boolean hasNext() {
+                storage.lock().readLock().lock();
+
+                try {
+                    while (true) {
+                        if (finished)
+                            return false;
+
+                        if (nextRetEntry != null)
+                            return true;
+
+                        byte[] key = lastRetKey;
+
+                        while (nextRetEntry == null) {
+                            Map.Entry<byte[], List<Long>> e =
+                                key == null ? storage.revisionCeilingEntry(keyFrom) : storage.revisionHigherEntry(key);
+
+                            if (e == null) {
+                                finished = true;
+
+                                break;
+                            }
+
+                            key = e.getKey();
+
+                            if (keyTo != null && RocksDBKeyValueStorage.CMP.compare(key, keyTo) >= 0) {
+                                finished = true;
+
+                                break;
+                            }
+
+                            List<Long> revs = e.getValue();
+
+                            assert revs != null && !revs.isEmpty() :
+                                "Revisions should not be empty or null: [revs=" + revs + ']';
+
+                            long lastRev = RocksDBKeyValueStorage.maxRevision(revs, rev);
+
+                            if (lastRev == -1)
+                                continue;
+
+                            Entry entry = storage.doGetValue(key, lastRev);
+
+                            assert !entry.empty() : "Iterator should not return empty entry.";
+
+                            nextRetEntry = entry;
+                        }
+                    }
+                }
+                finally {
+                    storage.lock().readLock().unlock();
+                }
+            }
+
+            /** {@inheritDoc} */
+            @Override public Entry next() {
+                storage.lock().readLock().lock();
+
+                try {
+                    if (!hasNext())
+                        throw new NoSuchElementException();
+
+                    Entry e = nextRetEntry;
+
+                    nextRetEntry = null;
+
+                    lastRetKey = e.key();
```

Comment:
Mine does not for some reason. One assertion won't hurt, of course, but I wonder how it can possibly be null if the `hasNext` method only returns `true` if it is not.